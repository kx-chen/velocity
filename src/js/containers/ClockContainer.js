import NameForm from "../components/NameForm";
import Time from "../components/Time";
import UnderClockMessage from "../components/UnderClockMessage";
import { determineMessageUnderClock, getCurrentTime } from "../utils/utils";
import { h } from "preact";
import { useEffect, useState } from "preact/hooks";
import { connect } from "react-redux";

export function ClockContainer({ timeFormat }) {
  const timePerTick = 500;
  // TODO: call getCurrentTime less

  const [minute, setMinute] = useState(getCurrentTime(timeFormat)[1]);
  const [hour, setHour] = useState(getCurrentTime(timeFormat)[0]);
  const [messageUnderClock, setMessageUnderClock] = useState(
    determineMessageUnderClock()
  );
  useEffect(() => {
    setMessageUnderClock(determineMessageUnderClock());
  }, [minute]);

  let interval1 = setInterval(
    () => setMinute(getCurrentTime(timeFormat)[1]),
    timePerTick
  );
  let interval2 = setInterval(
    () => setHour(getCurrentTime(timeFormat)[0]),
    timePerTick
  );

  useEffect(() => {
    // wait for next tick so that time updates before clearing intervals
    setTimeout(() => {
      clearInterval(interval1);
      clearInterval(interval2);
    }, 1000);
  });

  return (
    <div id="clock-wrapper">
      <NameForm />
      <Time hour={hour} minute={minute} />
      <UnderClockMessage message={messageUnderClock} />
    </div>
  );
}

function mapDispatchToProps(dispatch, ownProps) {
  return {
    dispatch,
  };
}

function mapStateToProps(state, ownProps) {
  return {
    ...state,
    timeFormat: state.time_format,
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(ClockContainer);
