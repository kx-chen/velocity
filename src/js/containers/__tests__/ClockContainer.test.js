import { ClockContainer } from "../ClockContainer";
import { h } from "preact";
import { mount } from "enzyme";

jest.mock("../../utils/utils", () => ({
  ...jest.requireActual("../../utils/utils"),
  getCurrentTime: jest
    .fn()
    // TODO: fix, hacky.
    .mockReturnValueOnce(["18", "59"])
    .mockReturnValueOnce(["18", "59"])
    .mockReturnValueOnce(["18", "59"])
    .mockReturnValueOnce(["18", "59"])
    .mockReturnValueOnce(["19", "59"])
    .mockReturnValueOnce(["19", "59"])
    .mockReturnValueOnce(["19", "59"])
    .mockReturnValueOnce(["19", "59"])
    .mockReturnValueOnce(["19", "59"])
    .mockReturnValueOnce(["19", "59"])
    .mockReturnValueOnce(["19", "59"])
    .mockReturnValueOnce(["19", "59"])
    .mockReturnValueOnce(["19", "59"])
    .mockReturnValueOnce(["19", "59"]),
  determineMessageUnderClock: jest.fn().mockReturnValue("Message Under Clock")
}));

describe("ClockContainer", () => {
  jest.useFakeTimers();

  test("updates clock and message when time changes", () => {
    const wrapper = mount(<ClockContainer />);

    expect(wrapper).toMatchSnapshot();
    jest.advanceTimersByTime(1000);
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
});
