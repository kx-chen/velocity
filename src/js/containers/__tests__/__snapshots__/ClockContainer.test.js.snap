// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClockContainer updates clock and message when time changes 1`] = `
<ClockContainer>
  <div
    id="clock-wrapper"
  >
    <NameForm>
      <div
        id="name-form"
      >
        <label
          className="name-label"
          htmlFor="name"
        >
          Hi. What's your name?
        </label>
        <input
          autoComplete="off"
          autoFocus={true}
          id="name"
          number="50"
          type="text"
        />
      </div>
    </NameForm>
    <Time
      hour="18"
      minute="59"
    >
      <div
        id="clock"
      >
        <span
          id="hour"
        >
          18
        </span>
        <span
          id="minute"
        >
          59
        </span>
      </div>
    </Time>
    <UnderClockMessage
      message="Message Under Clock"
    >
      <div
        id="message-under-clock"
      >
        Message Under Clock
      </div>
    </UnderClockMessage>
  </div>
</ClockContainer>
`;

exports[`ClockContainer updates clock and message when time changes 2`] = `
<ClockContainer>
  <div
    id="clock-wrapper"
  >
    <NameForm>
      <div
        id="name-form"
      >
        <label
          className="name-label"
          htmlFor="name"
        >
          Hi. What's your name?
        </label>
        <input
          autoComplete="off"
          autoFocus={true}
          id="name"
          number="50"
          type="text"
        />
      </div>
    </NameForm>
    <Time
      hour="19"
      minute="59"
    >
      <div
        id="clock"
      >
        <span
          id="hour"
        >
          19
        </span>
        <span
          id="minute"
        >
          59
        </span>
      </div>
    </Time>
    <UnderClockMessage
      message="Message Under Clock"
    >
      <div
        id="message-under-clock"
      >
        Message Under Clock
      </div>
    </UnderClockMessage>
  </div>
</ClockContainer>
`;
